frontend/src/metabase/visualizations/echarts/cartesian/model/axis.ts

```diff
@@ -0,0 +1,251 @@
+  seriesExtents: SeriesExtents,
+) {
+  const hasSingleCard =
+    seriesModels.reduce((cardIds, seriesModel) => {
+      cardIds.add(seriesModel.cardId);
+      return cardIds;
+    }, new Set<number | undefined>()).size === 1;
+
+  if (
+    settings["graph.y_axis.auto_split"] === false ||
+    (hasSingleCard && getAggregateColumnsCount(seriesModels) < 2) ||
+    settings["stackable.stack_type"] != null
+  ) {
+    return false;
+  }
+
+  const allMetricsColumns = seriesModels.map(seriesModel => seriesModel.column);
+
+  const hasDifferentYAxisColTypes =
+    _.uniq(allMetricsColumns.map(column => column.semantic_type)).length > 1;
+
+  if (hasDifferentYAxisColTypes) {
+    return true;
+  }
+
+  const allMetricsColumnSettings = allMetricsColumns
+    .map(column => settings.column?.(column))
+    .filter(isNotNull);
+
+  const columnSettings = allMetricsColumnSettings.map(columnSettings =>
+    getLineAreaBarComparisonSettings(columnSettings),
+  );
+
+  const hasDifferentColumnSettings = columnSettings.some(s1 =>
+    columnSettings.some(s2 => !_.isEqual(s1, s2)),
+  );
+
+  if (hasDifferentColumnSettings) {
+    return true;
+  }
+
+  const yExtents = Object.values(seriesExtents);
+
+  const minRange = Math.min(...yExtents.map(extent => extent[1] - extent[0]));
+  const maxExtent = Math.max(...yExtents.map(extent => extent[1]));
+  const minExtent = Math.min(...yExtents.map(extent => extent[0]));
+  const chartRange = maxExtent - minExtent;
+
+  // Note (EmmadUsmani): When the series with the smallest range is less than 5%
+  // of the chart's total range, we split the y-axis so it doesn't look too small.
+  return minRange / chartRange <= 0.05;
+}
+
+type AxisSplit = [DataKey[], DataKey[]];
+
+const SPLIT_AXIS_UNSPLIT_COST = -100;
+const SPLIT_AXIS_COST_FACTOR = 2;
+const SPLIT_AXIS_MAX_DEPTH = 8;
+
+function generateSplits(
+  list: DataKey[],
+  left: DataKey[] = [],
+  right: DataKey[] = [],
+  depth = 0,
+): AxisSplit[] {
+  // NOTE: currently generates all permutations, some of which are equivalent
+  if (list.length === 0) {
+    return [[left, right]];
+  } else if (depth > SPLIT_AXIS_MAX_DEPTH) {
+    // If we reach our max depth, we need to ensure that any item that haven't been added either list are accounted for
+    return left.length < right.length
+      ? [[left.concat(list), right]]
+      : [[left, right.concat(list)]];
+  } else {
+    return [
+      ...generateSplits(
+        list.slice(1),
+        left.concat([list[0]]),
+        right,
+        depth + 1,
+      ),
+      ...generateSplits(
+        list.slice(1),
+        left,
+        right.concat([list[0]]),
+        depth + 1,
+      ),
+    ];
+  }
+}
+
+function axisCost(extents: Extent[], favorUnsplit = true) {
+  const axisExtent = d3.extent(extents.flatMap(e => e));
+  const axisRange = axisExtent[1] - axisExtent[0];
+  if (favorUnsplit && extents.length === 0) {
+    return SPLIT_AXIS_UNSPLIT_COST;
+  } else if (axisRange === 0) {
```

Comment:
Nit: can just make this an `if` condition since it and the previous condition body both return, then the `else` on line 148 can be removed.